pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/LeaderService.java

```diff
@@ -0,0 +1,134 @@
+                         PulsarClient pulsarClient,
+                         FunctionRuntimeManager functionRuntimeManager,
+                         SchedulerManager schedulerManager,
+                         ErrorNotifier errorNotifier) {
+        this.workerConfig = workerService.getWorkerConfig();
+        this.pulsarClient = pulsarClient;
+        this.functionRuntimeManager = functionRuntimeManager;
+        this.schedulerManager = schedulerManager;
+        this.errorNotifier = errorNotifier;
+        consumerName = String.format(
+                "%s:%s:%d",
+                workerConfig.getWorkerId(),
+                workerConfig.getWorkerHostname(),
+                workerConfig.getWorkerPort()
+        );
+
+    }
+
+    public void start() throws PulsarClientException {
+        // the leaders service is using a `coordination` topic for leader election.
+        // we don't produce any messages into this topic, we only use the `failover` subscription
+        // to elect an active consumer as the leader worker. The leader worker will be responsible
+        // for scheduling snapshots for FMT and doing task assignment.
+        consumer = (ConsumerImpl<byte[]>) pulsarClient.newConsumer()
+                .topic(workerConfig.getClusterCoordinationTopic())
+                .subscriptionName(COORDINATION_TOPIC_SUBSCRIPTION)
+                .subscriptionType(SubscriptionType.Failover)
+                .consumerEventListener(this)
+                .property(WORKER_IDENTIFIER, consumerName)
+                .consumerName(consumerName)
+                .subscribe();
+
+    }
+
+    @Override
+    public void becameActive(Consumer<?> consumer, int partitionId) {
+        if (isLeader.compareAndSet(false, true)) {
+            log.info("Worker {} became the leader.", consumerName);
+            try {
+                // trigger read to the end of the topic and exit
+                // Since the leader can just update its in memory assignments cache directly
+                functionRuntimeManager.stopReadingAssignments();
+
+                // make sure scheduler is initialized because this worker
+                // is the leader and may need to start computing and writing assignments
+                schedulerManager.initialize();
+            } catch (Throwable th) {
+                log.error("Encountered error when initializing to become leader", th);
+                errorNotifier.triggerError(th);
+            }
+        }
+    }
+
+    @Override
+    public void becameInactive(Consumer<?> consumer, int partitionId) {
+        if (isLeader.compareAndSet(true, false)) {
+            log.info("Worker {} lost the leadership.", consumerName);
+            // when a worker has lost leadership it needs to start reading from the assignment topic again
+            try {
+                // acquire scheduler lock to make sure a scheduling is not in process
+                schedulerManager.getSchedulerLock().lock();
```

Comment:
I think the better way is to make scheduler aware of the leadership changes(just like runtime manger) and call acquireLeadership and giveupLeadership

Reply:
This is a way to do that.  You will need synchronization somewhere and someone will have to wait